server/coordinator/procedure/storage_impl.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"path"
+
+	"github.com/CeresDB/ceresmeta/server/etcdutil"
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+const (
+	Version              = "v1"
+	PathProcedure        = "procedure"
+	PathDeletedProcedure = "deletedProcedure"
+)
+
+type EtcdStorageImpl struct {
+	client    *clientv3.Client
+	clusterID uint32
+	rootPath  string
+}
+
+func NewEtcdStorageImpl(client *clientv3.Client, clusterID uint32, rootPath string) *EtcdStorageImpl {
+	return &EtcdStorageImpl{
+		client:    client,
+		clusterID: clusterID,
+		rootPath:  rootPath,
+	}
+}
+
+// CreateOrUpdate example:
+// procedure : /{rootPath}/v1/procedure/{clusterID}/{procedureID} -> {procedureType} + {procedureState} + {data}
+func (e EtcdStorageImpl) CreateOrUpdate(ctx context.Context, meta *Meta) error {
+	state := meta.State
+	if state == StateFinished || state == StateCancelled {
+		if err := e.deleteAndRecord(ctx, meta); err != nil {
+			return errors.WithMessage(err, "etcd delete data failed")
+		}
+		return nil
+	}
+
+	str, err := encode(meta)
+	if err != nil {
+		return errors.WithMessage(err, "encode meta failed")
+	}
+	keyPath := e.generateNormalKeyPath(meta.ID)
+	opPut := clientv3.OpPut(keyPath, str)
+
+	if _, err = e.client.Do(ctx, opPut); err != nil {
+		return errors.WithMessage(err, "etcd put data failed")
+	}
+	return nil
+}
+
+// Delete example:
+// historyProcedure : /{rootPath}/v1/historyProcedure/{clusterID}/{procedureID}
```

Comment:
Put all the comments as a statement as possible as you can:
```suggestion
// Do a soft deletion, and the deleted key's format is:
//  /{rootPath}/v1/historyProcedure/{clusterID}/{procedureID}
```